<a href="https://colab.research.google.com/github/Sagaust/DH-Computational-Methodologies/blob/main/Solution_3_%E2%80%93_Data_Loading_and_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### <b> Course: Machine Learning for Absolute Beginners - Level 2
#### By: Idan Gabrieli
####  Exercise #3 – Data Loading and Analysis
*****************************

### Step 1 - Loading a Tabular Data File

Make sure the "titanic.csv" file (provided as part of this excercise) is located in the JupyterLab working directory. <br> Read the "titanic.csv" file dataset into a dataframe called df. Print the first 5 rows in df.

In [ ]:
import pandas as pd
df = pd.read_csv('titanic.csv', header=0)
df.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

Read again the "titanic.csv" file dataset into a dataframe called df but without the following columns: "SibSp", "Parch" and "Ticket". Print the first 5 rows in df.

In [ ]:
df = pd.read_csv('titanic.csv', header=0, usecols = [0,1,2,3,4,5,9,10,11])
df.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age     Fare  \
0                            Braund, Mr. Owen Harris    male  22.0   7.2500   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0  71.2833   
2                             Heikkinen, Miss. Laina  female  26.0   7.9250   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0  53.1000   
4                           Allen, Mr. William Henry    male  35.0   8.0500   

  Cabin Embarked  
0   NaN        S  
1   C85        C  
2   NaN        S  
3  C123        S  
4   NaN        S

Print the data type of each column by using the dtypes attribute.

In [ ]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
Fare           float64
Cabin           object
Embarked        object
dtype: object

Read again the "titanic.csv" file dataset into a dataframe called df but without the following columns: "SibSp", "Parch" and "Ticket" and also convert the column "Survived" into a boolean data type. Print the first 5 rows in df.

In [ ]:
df = pd.read_csv('titanic.csv', header=0, usecols = [0,1,2,3,4,5,9,10,11], dtype = {'Survived':'bool'})
df.head(5)

PassengerId  Survived  Pclass  \
0            1     False       3   
1            2      True       1   
2            3      True       3   
3            4      True       1   
4            5     False       3   

                                                Name     Sex   Age     Fare  \
0                            Braund, Mr. Owen Harris    male  22.0   7.2500   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0  71.2833   
2                             Heikkinen, Miss. Laina  female  26.0   7.9250   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0  53.1000   
4                           Allen, Mr. William Henry    male  35.0   8.0500   

  Cabin Embarked  
0   NaN        S  
1   C85        C  
2   NaN        S  
3  C123        S  
4   NaN        S

Print the data type of each column by using the dtypes attribute. The "Survived" should be boolean data type.

In [ ]:
df.dtypes

PassengerId      int64
Survived          bool
Pclass           int64
Name            object
Sex             object
Age            float64
Fare           float64
Cabin           object
Embarked        object
dtype: object

### Step 2 - Preview the DataFrame

Print the two index components of the df dataframe: index, columns.

In [ ]:
print(df.index)
print(df.columns)

RangeIndex(start=0, stop=891, step=1)
Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'Fare',
       'Cabin', 'Embarked'],
      dtype='object')


Print the size of the dataframe using the shape attribute. We have 891 rows and 9 columns.

In [ ]:
df.shape

(891, 9)

Print the amount of rows per each column without missing values.

In [ ]:
df.count()

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
Fare           891
Cabin          204
Embarked       889
dtype: int64

Print the unique values for the "Sex" column.

In [ ]:
df['Sex'].unique()

array(['male', 'female'], dtype=object)

Print the unique values for the "Embarked" column.

In [ ]:
df['Embarked'].unique()

array(['S', 'C', 'Q', nan], dtype=object)

Print the amount of unique values for the "Embarked" column.

In [ ]:
len(df['Embarked'].unique())

4

Print the number of rows with the same specific value in the "Embarked" column.

In [ ]:
df['Embarked'].value_counts()

S    644
C    168
Q     77
Name: Embarked, dtype: int64

How many passengers survived?

In [ ]:
df['Survived'].value_counts()

False    549
True     342
Name: Survived, dtype: int64

What is the survival percentage?

In [ ]:
df['Survived'].value_counts(normalize=True)

False    0.616162
True     0.383838
Name: Survived, dtype: float64

### Step 3 - Using Summary Statistics

What is the maximum age of the titanic passengers?

In [ ]:
df["Age"].max()

80.0

What is the average age and average ticket price?

In [ ]:
df[["Age", "Fare"]].mean()

Age     29.699118
Fare    32.204208
dtype: float64

How many passengers survived? (using the sum method).

In [ ]:
df['Survived'].sum()

342

Print a full statistical summary using the describe method. What is the agev value that 25% of passengers are smaller then this age? (should be 20.123).

In [ ]:
df.describe()

PassengerId      Pclass         Age        Fare
count   891.000000  891.000000  714.000000  891.000000
mean    446.000000    2.308642   29.699118   32.204208
std     257.353842    0.836071   14.526497   49.693429
min       1.000000    1.000000    0.420000    0.000000
25%     223.500000    2.000000   20.125000    7.910400
50%     446.000000    3.000000   28.000000   14.454200
75%     668.500000    3.000000   38.000000   31.000000
max     891.000000    3.000000   80.000000  512.329200

### Step 4 - Sorting and Ranking

Sort all rows in the dataframe using the 'Age' column is descending order and print the first 5 rows.

In [ ]:
df.sort_values('Age', ascending=False).head(5)

PassengerId  Survived  Pclass                                  Name  \
630          631      True       1  Barkworth, Mr. Algernon Henry Wilson   
851          852     False       3                   Svensson, Mr. Johan   
493          494     False       1               Artagaveytia, Mr. Ramon   
96            97     False       1             Goldschmidt, Mr. George B   
116          117     False       3                  Connors, Mr. Patrick   

      Sex   Age     Fare Cabin Embarked  
630  male  80.0  30.0000   A23        S  
851  male  74.0   7.7750   NaN        S  
493  male  71.0  49.5042   NaN        C  
96   male  71.0  34.6542    A5        C  
116  male  70.5   7.7500   NaN        Q

Sort all rows in the dataframe using the 'Age' column is descending order and save it in the same dataframe. Print the first 5 rows.

In [ ]:
df.sort_values('Age', ascending=False, inplace=True)
df.head(5)

PassengerId  Survived  Pclass                                  Name  \
630          631      True       1  Barkworth, Mr. Algernon Henry Wilson   
851          852     False       3                   Svensson, Mr. Johan   
493          494     False       1               Artagaveytia, Mr. Ramon   
96            97     False       1             Goldschmidt, Mr. George B   
116          117     False       3                  Connors, Mr. Patrick   

      Sex   Age     Fare Cabin Embarked  
630  male  80.0  30.0000   A23        S  
851  male  74.0   7.7750   NaN        S  
493  male  71.0  49.5042   NaN        C  
96   male  71.0  34.6542    A5        C  
116  male  70.5   7.7500   NaN        Q

Sort the data based on male and female groups (female group is first) and inside of each group, sort the rows by age in ascending order. Present just the following columns 'Sex', 'Name', 'Age'.

In [ ]:
df.sort_values(['Sex', 'Age'], ascending=[True, True])[[ 'Sex', 'Name', 'Age']]

Sex                           Name   Age
644  female         Baclini, Miss. Eugenie  0.75
469  female  Baclini, Miss. Helene Barbara  0.75
381  female    Nakid, Miss. Maria ("Mary")  1.00
172  female   Johnson, Miss. Eleanor Ileen  1.00
642  female  Skoog, Miss. Margit Elizabeth  2.00
..      ...                            ...   ...
839    male           Marechal, Mr. Pierre   NaN
846    male       Sage, Mr. Douglas Bullen   NaN
859    male               Razi, Mr. Raihed   NaN
868    male    van Melkebeke, Mr. Philemon   NaN
878    male             Laleff, Mr. Kristo   NaN

[891 rows x 3 columns]

Print the row in location 2 using the iloc method.

Drop the missing values and then sort the data based on male and female groups (female group is first) and inside of each group, sort the rows by age in ascending order. Present just the following columns 'Sex', 'Name', 'Age'.

In [ ]:
df.dropna().sort_values(['Sex', 'Age'], ascending=[True, True])[[ 'Sex', 'Name', 'Age']]

Sex                                  Name   Age
205  female            Strom, Miss. Telma Matilda   2.0
297  female          Allison, Miss. Helen Loraine   2.0
10   female       Sandstrom, Miss. Marguerite Rut   4.0
618  female           Becker, Miss. Marion Louise   4.0
435  female             Carter, Miss. Lucile Polk  14.0
..      ...                                   ...   ...
54     male        Ostby, Mr. Engelhart Cornelius  65.0
456    male             Millet, Mr. Francis Davis  65.0
745    male          Crosby, Capt. Edward Gifford  70.0
96     male             Goldschmidt, Mr. George B  71.0
630    male  Barkworth, Mr. Algernon Henry Wilson  80.0

[183 rows x 3 columns]

What are the five oldest passengers that paid a ticket that cost more than the average ticket price? (create a boolean condition and than use it to filter the dataframe and than use the nlargest.

In [ ]:
filter=df['Fare']>df['Fare'].mean()
df[filter].nlargest(5, 'Age')[['Name', 'Age', 'Fare']]

Name   Age      Fare
493         Artagaveytia, Mr. Ramon  71.0   49.5042
96        Goldschmidt, Mr. George B  71.0   34.6542
745    Crosby, Capt. Edward Gifford  70.0   71.0000
54   Ostby, Mr. Engelhart Cornelius  65.0   61.9792
438               Fortune, Mr. Mark  64.0  263.0000

### Step 5 - Filtering

Filter and keep only the male passengers.

In [ ]:
df[df['Sex']=='male']

PassengerId  Survived  Pclass                                  Name  \
630          631      True       1  Barkworth, Mr. Algernon Henry Wilson   
851          852     False       3                   Svensson, Mr. Johan   
493          494     False       1               Artagaveytia, Mr. Ramon   
96            97     False       1             Goldschmidt, Mr. George B   
116          117     False       3                  Connors, Mr. Patrick   
..           ...       ...     ...                                   ...   
839          840      True       1                  Marechal, Mr. Pierre   
846          847     False       3              Sage, Mr. Douglas Bullen   
859          860     False       3                      Razi, Mr. Raihed   
868          869     False       3           van Melkebeke, Mr. Philemon   
878          879     False       3                    Laleff, Mr. Kristo   

      Sex   Age     Fare Cabin Embarked  
630  male  80.0  30.0000   A23        S  
851  male  74.0   7.7750   NaN        S  
493  male  71.0  49.5042   NaN        C  
96   male  71.0  34.6542    A5        C  
116  male  70.5   7.7500   NaN        Q  
..    ...   ...      ...   ...      ...  
839  male   NaN  29.7000   C47        C  
846  male   NaN  69.5500   NaN        S  
859  male   NaN   7.2292   NaN        C  
868  male   NaN   9.5000   NaN        S  
878  male   NaN   7.8958   NaN        S  

[577 rows x 9 columns]

Filer out all passengers below the age of 40 and that are also 'male'. Print the ten first results.

In [ ]:
df[(df['Age']>=40) & (df['Sex'] !='male')].head(10)

PassengerId  Survived  Pclass  \
275          276      True       1   
483          484      True       3   
829          830      True       1   
366          367      True       1   
268          269      True       1   
11            12      True       1   
195          196      True       1   
772          773     False       2   
879          880      True       1   
15            16      True       2   

                                                 Name     Sex   Age      Fare  \
275                 Andrews, Miss. Kornelia Theodosia  female  63.0   77.9583   
483                            Turkula, Mrs. (Hedwig)  female  63.0    9.5875   
829         Stone, Mrs. George Nelson (Martha Evelyn)  female  62.0   80.0000   
366  Warren, Mrs. Frank Manley (Anna Sophia Atkinson)  female  60.0   75.2500   
268     Graham, Mrs. William Thompson (Edith Junkins)  female  58.0  153.4625   
11                           Bonnell, Miss. Elizabeth  female  58.0   26.5500   
195                              Lurette, Miss. Elise  female  58.0  146.5208   
772                                 Mack, Mrs. (Mary)  female  57.0   10.5000   
879     Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0   83.1583   
15                   Hewlett, Mrs. (Mary D Kingcome)   female  55.0   16.0000   

    Cabin Embarked  
275    D7        S  
483   NaN        S  
829   B28      NaN  
366   D37        C  
268  C125        S  
11   C103        S  
195   B80        C  
772   E77        S  
879   C50        C  
15    NaN        S

### Step 6 - Grouping and Aggregating

What is the average ticket price for male versus female passengers?

In [ ]:
df[["Sex", "Fare"]].groupby("Sex").mean()

Fare
Sex              
female  44.479818
male    25.523893

What is the average age for the survived group?

In [ ]:
df.groupby('Survived').mean()

PassengerId    Pclass        Age       Fare
Survived                                             
False      447.016393  2.531876  30.626179  22.117887
True       444.368421  1.950292  28.343690  48.395408

How many survived group by the embarked port?

In [ ]:
df[['Embarked','Survived']].groupby('Embarked').sum()

Survived
Embarked          
C               93
Q               30
S              217

What is the average age as well as the average price for male and female?

In [ ]:
df.groupby('Sex').mean()

PassengerId  Survived    Pclass        Age       Fare
Sex                                                          
female   431.028662  0.742038  2.159236  27.915709  44.479818
male     454.147314  0.188908  2.389948  30.726645  25.523893

### End of document
************************